File: client-scalaz/src/main/scala/org/openapitools/client/api/AssignmentResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import AssignmentResponse._

case class AssignmentResponse (
  assignmentId: Option[Long],
name: Option[String],
description: Option[String],
active: Option[Boolean],
created: Option[Long],
updated: Option[Long],
assignee: Option[AccountShortResponse],
creator: Option[AccountShortResponse],
lastReportedBy: Option[AccountShortResponse],
location: Option[RetailerLocationShortResponse],
currentStatus: Option[AssignmentStatusResponse],
currentStatusType: Option[CurrentStatusType])

object AssignmentResponse {
  import DateTimeCodecs._
  sealed trait CurrentStatusType
  case object `NEW` extends CurrentStatusType
  case object INPROGRESS extends CurrentStatusType
  case object SUBSCRIBED extends CurrentStatusType
  case object ARCHIVED extends CurrentStatusType

  object CurrentStatusType {
    def toCurrentStatusType(s: String): Option[CurrentStatusType] = s match {
      case "`NEW`" => Some(`NEW`)
      case "INPROGRESS" => Some(INPROGRESS)
      case "SUBSCRIBED" => Some(SUBSCRIBED)
      case "ARCHIVED" => Some(ARCHIVED)
      case _ => None
    }

    def fromCurrentStatusType(x: CurrentStatusType): String = x match {
      case `NEW` => "`NEW`"
      case INPROGRESS => "INPROGRESS"
      case SUBSCRIBED => "SUBSCRIBED"
      case ARCHIVED => "ARCHIVED"
    }
  }

  implicit val CurrentStatusTypeEnumEncoder: EncodeJson[CurrentStatusType] =
    EncodeJson[CurrentStatusType](is => StringEncodeJson(CurrentStatusType.fromCurrentStatusType(is)))

  implicit val CurrentStatusTypeEnumDecoder: DecodeJson[CurrentStatusType] =
    DecodeJson.optionDecoder[CurrentStatusType](n => n.string.flatMap(jStr => CurrentStatusType.toCurrentStatusType(jStr)), "CurrentStatusType failed to de-serialize")

  implicit val AssignmentResponseCodecJson: CodecJson[AssignmentResponse] = CodecJson.derive[AssignmentResponse]
  implicit val AssignmentResponseDecoder: EntityDecoder[AssignmentResponse] = jsonOf[AssignmentResponse]
  implicit val AssignmentResponseEncoder: EntityEncoder[AssignmentResponse] = jsonEncoderOf[AssignmentResponse]
}

File: client-clojure/src/sirqul_io_t_platform/specs/assignment_response.clj
(ns sirqul-io-t-platform.specs.assignment-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account-short-response :refer :all]
            [sirqul-io-t-platform.specs.account-short-response :refer :all]
            [sirqul-io-t-platform.specs.account-short-response :refer :all]
            [sirqul-io-t-platform.specs.retailer-location-short-response :refer :all]
            [sirqul-io-t-platform.specs.assignment-status-response :refer :all]
            )
  (:import (java.io File)))


(def assignment-response-data
  {
   (ds/opt :assignmentId) int?
   (ds/opt :name) string?
   (ds/opt :description) string?
   (ds/opt :active) boolean?
   (ds/opt :created) int?
   (ds/opt :updated) int?
   (ds/opt :assignee) account-short-response-spec
   (ds/opt :creator) account-short-response-spec
   (ds/opt :lastReportedBy) account-short-response-spec
   (ds/opt :location) retailer-location-short-response-spec
   (ds/opt :currentStatus) assignment-status-response-spec
   (ds/opt :currentStatusType) string?
   })

(def assignment-response-spec
  (ds/spec
    {:name ::assignment-response
     :spec assignment-response-data}))
